File: ios/RCTMGL-v10/RCTMGLShapeSourceManager.m
#import <React/RCTBridgeModule.h>
#import <React/RCTViewManager.h>

@interface
RCT_EXTERN_REMAP_MODULE(RCTMGLShapeSource, RCTMGLShapeSourceManager, RCTViewManager)

RCT_EXPORT_VIEW_PROPERTY(id, NSString)
RCT_EXPORT_VIEW_PROPERTY(existing, BOOL)
RCT_EXPORT_VIEW_PROPERTY(url, NSString)
RCT_EXPORT_VIEW_PROPERTY(shape, NSString)
RCT_EXPORT_VIEW_PROPERTY(lineStartOffset, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(lineEndOffset, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(animationDuration, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(snapIfDistanceIsGreaterThan, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(cluster, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(clusterRadius, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(clusterMaxZoomLevel, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(clusterProperties, NSDictionary)
RCT_EXPORT_VIEW_PROPERTY(maxZoomLevel, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(buffer, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(tolerance, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(lineMetrics, NSNumber)
RCT_EXPORT_VIEW_PROPERTY(images, NSDictionary)
RCT_EXPORT_VIEW_PROPERTY(nativeImages, NSArray)
RCT_EXPORT_VIEW_PROPERTY(hasPressListener, BOOL)
RCT_EXPORT_VIEW_PROPERTY(hitbox, NSDictionary)
RCT_REMAP_VIEW_PROPERTY(onMapboxShapeSourcePress, onPress, RCTBubblingEventBlock)

RCT_EXTERN_METHOD(getClusterExpansionZoom:(nonnull NSNumber*)reactTag
                                featureJSON:(nonnull NSString*)featureJSON
                                resolver:(RCTPromiseResolveBlock)resolve
                                rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getClusterLeaves:(nonnull NSNumber*)reactTag
                  featureJSON:(nonnull NSString*)featureJSON
                  number:(NSUInteger) number
                  offset:(NSUInteger) offset
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getClusterChildren:(nonnull NSNumber*)reactTag
                  featureJSON:(nonnull NSString*)featureJSON
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/RCTMGL-v10/RCTMGLShapeSourceManager.swift
@objc(RCTMGLShapeSourceManager)
class RCTMGLShapeSourceManager: RCTViewManager {
  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc override func view() -> UIView {
    return RCTMGLShapeSource()
  }
}

// MARK: - helpers

extension RCTMGLShapeSourceManager {
  func withShapeSource(
    _ reactTag: NSNumber,
    name: String,
    rejecter: @escaping RCTPromiseRejectBlock,
    fn: @escaping (_: RCTMGLShapeSource) -> Void) {
    self.bridge.uiManager.addUIBlock { (_, viewRegistry) in
      let view = viewRegistry![reactTag]

      guard let shapeSource = view! as? RCTMGLShapeSource else {
        RCTMGLLogError("Invalid react tag, could not find RCTMGLShapeSource")
        rejecter(name, "Unknown find reactTag: \(reactTag)", nil)
        return
      }

      fn(shapeSource)
    }
  }
}

// MARK: - react methods

extension RCTMGLShapeSourceManager {
  @objc func getClusterExpansionZoom(
    _ reactTag: NSNumber,
    featureJSON: String,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock) {
    self.withShapeSource(reactTag, name: "getCluster   ExpansionZoom", rejecter: rejecter) { shapeSource in
      shapeSource.getClusterExpansionZoom(featureJSON) { result in
        switch result {
        case .success(let zoom):
          resolver([
            "data": NSNumber(value: zoom)
          ])

        case .failure(let error):
          rejecter(error.localizedDescription, "Error.getClusterExpansionZoom", error)
        }
      }
    }
  }

  @objc func getClusterLeaves(
    _ reactTag: NSNumber,
    featureJSON: String,
    number: uint,
    offset: uint,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock) {
    self.withShapeSource(reactTag, name: "getClusterLeaves", rejecter: rejecter) { shapeSource in
      shapeSource.getClusterLeaves(featureJSON, number: number, offset: offset) { result in
        switch result {
        case .success(let features):
          logged("getClusterLeaves", rejecter: rejecter) {
            let featuresJSON: Any = try features.features.toJSON()
            resolver([
              "data": ["type": "FeatureCollection", "features": featuresJSON]
            ])
          }

        case .failure(let error):
          rejecter(error.localizedDescription, "Error.getClusterLeaves", error)
        }
      }
    }
  }

  @objc func getClusterChildren(
    _ reactTag: NSNumber,
    featureJSON: String,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock) {
    self.withShapeSource(reactTag, name: "getClusterChildren", rejecter: rejecter) { shapeSource in
      shapeSource.getClusterChildren(featureJSON) { result in
        switch result {
        case .success(let features):
          logged("getClusterChildren", rejecter: rejecter) {
            let featuresJSON: Any = try features.features.toJSON()
            resolver([
              "data": ["type": "FeatureCollection", "features": featuresJSON]
            ])
          }

        case .failure(let error):
          rejecter(error.localizedDescription, "Error.getClusterChildren", error)
        }
      }
    }
  }
}
